File: backend/app/api/routes/workshop_types.py
"""``/api/workshop-types`` — the administrator-managed "Type of workshop" list.

WHAT THIS IS, IN ONE PARAGRAPH. Every record form in this product (artisan, product, process, tool,
craft, questionnaire) is about to carry exactly TWO workshop controls: "Type of workshop", filled
from this endpoint, and "Workshop", filled from whichever table the chosen type points at. This
router owns the first list and the flag that decides the second. It owns nothing else — no record is
written here, no workshop is created here, and the type a researcher picks is not stored on the
record they save.

── THE PERMISSION SPLIT, AND IT IS THE ONE `crafts.py` USES ───────────────────────────────────────

READ is ``get_current_user``: every signed-in account needs this list, because every record form
draws it. WRITE is ``require_admin`` — MASTER_ADMIN and ADMIN, the same helper and the same set as
``/admin`` and every other admin-only surface — because this is a shared vocabulary and renaming a
type renames it under everybody. That asymmetry is not an oversight in either direction: a list
nobody could read would be a dropdown with no members, and a list anybody could write would let a
researcher rename the category another researcher's fieldwork is filed under.

The gate is HERE, on the route, and not only on the admin screen. A designer who types
``/admin/workshop-types`` meets ``AppShell``'s lock panel; a designer who types
``POST /api/workshop-types`` meets a 403 from ``require_admin``. The second one is the boundary; the
first is a courtesy.

── WHY DELETING IS REFUSED RATHER THAN CASCADED, AND WHAT IS OFFERED INSTEAD ──────────────────────

**THERE IS NO FOREIGN KEY ONTO ``WorkshopTypeOption``**, by design — a record stores the WORKSHOP it
belongs to and the workshop already knows its own type, so a second copy of the type on the record
would be a second copy that can disagree. The cost of that decision lands exactly here: Postgres
cannot refuse this DELETE for us, because from its point of view nothing references the row.

So the refusal is written by hand, and it is a REFUSAL rather than a cascade or a silent success:

  * A type with nothing filed under it is deleted, 204, no ceremony. This is the ordinary case for a
    type an administrator added by mistake ten seconds ago, and making that a two-step retirement
    would leave the list carrying somebody's typo forever.
  * A type with workshops under it is refused 409, and the sentence NAMES HOW MANY and names the
    remedy. Deactivating (``PATCH {"isActive": false}``) takes the type out of every picker while
    leaving every workshop filed under it exactly where it is, with its token still resolving to a
    label — which is the whole point, because the token is stored in `DesignWorkshop.workshopKind`,
    in `AnnualPlanEntry.workshopKind`, inside stage documents and in every export ever taken, and
    none of those is reachable from here.

The alternative — delete the row and let the workshops keep a token nothing can resolve — is the
failure this repository has shipped often enough to have a name for it: the workshop would render
with no type at all, indistinguishable from a workshop whose type was never answered, and the only
record that the category ever existed would be in an export somebody took last year.

── EDITING A LABEL IS ALWAYS SAFE, AND THE SCHEMA IS WHAT GUARANTEES IT ───────────────────────────

``WorkshopTypeOptionUpdate`` does not carry ``key`` at all, and ``APIModel`` is ``extra="forbid"``,
so a PATCH that tries to move a key is answered 422 rather than quietly ignored. See that module's
header for the four places the token is already written and why none of them is rewritable from
here. A label edit touches one column and can never be anything else.
"""

from __future__ import annotations

import logging
from typing import Any, NamedTuple

from fastapi import APIRouter, Depends, HTTPException, Query, status
from prisma.errors import UniqueViolationError

from app.core.db import db
from app.core.deps import get_current_user, require_admin
from app.schemas.workshop_types import (
    WorkshopTypeOptionCreate,
    WorkshopTypeOptionReorder,
    WorkshopTypeOptionUpdate,
)
from app.services.records import public_encode

logger = logging.getLogger(__name__)

router = APIRouter(prefix="/workshop-types", tags=["workshop-types"])

#: The order every reader sees, and it is TOTAL. ``sortOrder`` is not unique (the model comment says
#: why a unique ordinal is refused), so ``key`` breaks the tie — without it two identical requests
#: could legitimately return two different lists, and a dropdown that reshuffles between page loads
#: is one a researcher stops trusting.
_ORDER: list[dict[str, str]] = [{"sortOrder": "asc"}, {"key": "asc"}]


class Noun(NamedTuple):
    """What a count is counting, in both numbers, as the refusal will print it.

    BOTH SPELLINGS ARE DECLARED AND NEITHER IS DERIVED. English pluralises the HEAD of a noun
    phrase, not its end: "design workshop" takes an "s" and "planned workshop in the annual
    directory" does not — "planned workshopS in the annual directory" is correct and
    "planned workshop in the annual directorys" is what a bare rule produces. A refusal an
    administrator reads at the moment a delete fails is the last place to print that.
    """

    one: str
    many: str


#: WHAT COUNTS AS "IN USE", AS ``(prisma model attribute, column, :class:`Noun`, extra where)``.
#:
#: THE FOURTH MEMBER IS WHAT "STILL COUNTS" MEANS FOR THAT TABLE, declared beside the table rather
#: than as a branch inside :func:`_count_use` on the model's name. The two rows disagree about it and
#: the disagreement is deliberate:
#:
#:   * ``DesignWorkshop`` narrows on ``deletedAt: None``. A workshop in the trash is not in use.
#:     Nothing here is ever hard-deleted and every read in the product filters that column, so
#:     counting the trash would refuse the delete of a type whose only workshops are in the bin — and
#:     nothing would be able to tell the administrator WHICH, because the only screen that lists them
#:     is the admin hub's recovery card.
#:   * ``AnnualPlanEntry`` narrows on nothing. Its retirement column is ``withdrawnAt``, and a
#:     WITHDRAWN plan row is still counted on purpose: the directory keeps it on screen, still
#:     rendering its kind, precisely because the plan is the record of what WAS intended.
#:
#: The shape — and the reason a missing model answers ``None`` rather than raising — is
#: ``routes/users.py::_NAMED_ON_RELATIONS`` and ``_count_relation``, whose docstrings argue it at
#: length and are worth reading before touching this: a generated Prisma client that has drifted
#: behind ``schema.prisma`` is the DEFAULT outcome of running ``prisma generate`` on Windows (it
#: prints ``Error: spawn prisma-client-py ENOENT`` and exits 0), so code downstream of it cannot
#: treat the client's contents as a guarantee. Here the consequence is milder than it is there —
#: this router's refusal is a decision rather than a decoration on one Postgres has already made —
#: which is exactly why the drift must not be allowed to turn it into a 500. An UNCOUNTABLE relation
#: refuses the delete too, and says it could not count; see :func:`_usage_tally`.
#:
#: ``AnnualPlanEntry`` IS ON THIS LIST AND IS DELIBERATELY NOT CALLED A WORKSHOP. ``api/router.py``
#: is emphatic that a row on ``/annual-plan`` "is a line in a document and not a workshop", and
#: counting plan rows into a workshop total would be the exact confusion that file exists to stop.
#: It is here because it stores the same token in the same spelling, so deleting a type would leave
#: the ministry's directory rendering a kind nothing can resolve — the same failure, on a different
#: screen — and it is reported under its own noun so an administrator can see which is which.
_USES: tuple[tuple[str, str, Noun, dict[str, Any]], ...] = (
    (
        "designworkshop",
        "workshopKind",
        Noun("design workshop", "design workshops"),
        {"deletedAt": None},
    ),
    (
        "annualplanentry",
        "workshopKind",
        Noun(
            "planned workshop in the annual directory",
            "planned workshops in the annual directory",
        ),
        {},
    ),
)


def _payload(row: Any) -> dict[str, Any]:
    """One row, as every client reads it.

    ``public_encode`` rather than a hand-built dict, for the reason every other router uses it: it is
    the one function that knows what must never leave this API. Nothing on this table is sensitive
    today, and a payload built by hand here is a payload that would not be scrubbed on the day
    somebody adds a column that is.
    """
    return public_encode(row)


async def _count_use(
    client: Any, model: str, column: str, key: str, extra: dict[str, Any]
) -> int | None:
    """How many rows of ``model`` are filed under ``key``, or ``None`` if the count could not be run.

    ``client`` IS AN ARGUMENT AND NOT THE MODULE SINGLETON, and that is not decoration. ``db.tx()``
    hands back a DIFFERENT client, so a callee reading through ``db`` is NOT inside the caller's
    transaction however it looks — ``crafts.py::update_craft`` writes that warning out in full where
    it passes ``client=tx``. The count that decides this delete has to run in the same transaction as
    the delete, so the client travels down rather than being reached for.

    ``None`` RATHER THAN ``0``, and the distinction reaches the administrator's screen. A relation
    that could not be read is not a relation with no rows; answering 0 would let this router delete a
    type that twelve workshops are filed under, because a migration had not been applied to the
    database serving the request. ``users.py::_count_relation`` argues the same distinction for a
    message; here it decides whether a row is destroyed, so the safe direction is the opposite of
    that one's — see :func:`_usage_tally`.
    """
    delegate = getattr(client, model, None)
    if delegate is None:
        logger.error(
            "workshop_types: the generated Prisma client carries no model %r, so a delete of a "
            "workshop type cannot check whether %s.%s still uses it. The client is behind "
            "schema.prisma — regenerate it with backend/scripts/regenerate-client.md, NOT with a "
            "bare `prisma generate`, which exits 0 without doing anything on Windows. Refusing the "
            "delete rather than performing one that cannot be checked; see _count_use.",
            model,
            model,
            column,
        )
        return None
    try:
        # ``extra`` is declared beside its table in :data:`_USES`, which is where the argument for
        # each narrowing lives. It is spread AFTER the key so a malformed entry can never shadow the
        # thing being counted.
        return await delegate.count(where={column: key, **extra})
    except Exception:
        # Deliberately every exception and not a named Prisma error, for the reason
        # ``users.py::_count_relation`` gives: the set of ways a count can fail — a migration not
        # applied to this database, a column applied in part, the pool exhausted, the connection
        # dropped — is not a set this file can enumerate correctly and has no business trying to.
        # ``CancelledError`` derives from ``BaseException``, so a cancelled request still cancels.
        logger.exception(
            "workshop_types: counting %s.%s for workshop type key %r failed, so a delete cannot be "
            "checked. Usually a migration that has not been applied to this database. Refusing the "
            "delete; see _count_use.",
            model,
            column,
            key,
        )
        return None


async def _usage_tally(client: Any, key: str) -> tuple[list[tuple[Noun, int]], list[Noun]]:
    """``([(noun, count)], [uncountable noun])`` for one key — what stands in the way of a delete.

    TWO LISTS RATHER THAN ONE, because they say different things and an administrator must not be
    able to confuse them: the first is "these are in the way, this many of them", the second is "and
    I could not look here at all". Collapsing an unreadable relation into a zero would be a lie of
    exactly the kind this function exists to stop telling — and, unlike the tally in ``users.py``
    which decorates a refusal Postgres has already made, THIS one decides whether rows are destroyed.
    So an uncountable relation refuses the delete rather than being skipped.
    """
    named: list[tuple[Noun, int]] = []
    uncounted: list[Noun] = []
    for model, column, noun, extra in _USES:
        count = await _count_use(client, model, column, key, extra)
        if count is None:
            uncounted.append(noun)
        elif count > 0:
            named.append((noun, count))
    named.sort(key=lambda pair: pair[1], reverse=True)
    return named, uncounted


#: The remedy, and the half of the refusal that decides whether an administrator gets what they
#: actually wanted. One constant because all three branches of :func:`_in_use_message` end with it.
_DEACTIVATE_INSTEAD = (
    " Deactivate it instead — it disappears from every “Type of workshop” dropdown and every "
    "workshop already filed under it keeps its type and its label."
)


def _in_use_message(label: str, named: list[tuple[Noun, int]], uncounted: list[Noun]) -> str:
    """The 409, worded for the administrator who pressed Delete and has to decide what to do next.

    IT NAMES THE NUMBER, THE THING COUNTED AND THE REMEDY, in that order, because those are the three
    things missing from "Cannot delete: in use". The remedy is the sentence that matters most: an
    administrator who is told only that they may not delete something will try again tomorrow, and
    "deactivate it instead" is an action available on the same screen that does what they actually
    wanted — the type disappears from every dropdown — without touching a single workshop.

    THREE BRANCHES, BECAUSE "I COUNTED SOME" AND "I COULD NOT LOOK" ARE DIFFERENT SENTENCES AND THE
    SECOND ONE ALONE HAS TO STAND UP ON ITS OWN. Gluing the uncounted clause onto the end of a
    comma-list is how a refusal comes to read "“Other” cannot be deleted: and this server could not
    check … still filed under it" — a sentence that is about nothing and is the ONLY thing on screen
    at the moment an administrator most needs to understand what happened.

    THE PLURAL COMES FROM :data:`_USES` AND IS NOT DERIVED. A bare "s" is right for "design
    workshop" and wrong for "planned workshop in the annual directory" — English pluralises the HEAD
    of a noun phrase, not its end — so the two spellings are declared beside the table and this
    function only chooses between them.
    """
    if named:
        counted = ", ".join(
            f"{count} {noun.one if count == 1 else noun.many}" for noun, count in named
        )
        sentence = f"“{label}” cannot be deleted: {counted} still filed under it."
        if uncounted:
            # Worded as a fact about this SERVER rather than about the data, so nobody reads it as
            # "there might be some": the count did not run, and that is a deployment problem with a
            # log line behind it.
            sentence += (
                " This server could also not check "
                + ", ".join(noun.many for noun in uncounted)
                + ", so there may be more."
            )
        return sentence + _DEACTIVATE_INSTEAD
    return (
        f"“{label}” cannot be deleted: this server could not check whether any "
        + ", ".join(noun.many for noun in uncounted)
        + " are filed under it, so the delete was refused rather than performed unchecked. The "
        "cause is on this server rather than in the data — usually a migration that has not been "
        "applied here, or a Prisma client behind schema.prisma — and it is in the logs."
        + _DEACTIVATE_INSTEAD
    )


async def _require_type(type_id: str) -> Any:
    row = await db.workshoptypeoption.find_unique(where={"id": type_id})
    if row is None:
        raise HTTPException(status_code=status.HTTP_404_NOT_FOUND, detail="Workshop type not found")
    return row


@router.get("")
async def list_workshop_types(
    _: Any = Depends(get_current_user),
    # DEFAULT FALSE, so the picker gets the list it should offer without asking for it. A retired
    # type must not appear in a dropdown — that is the whole meaning of retiring one — and a default
    # that included them would put the decision in every call site instead of in one.
    includeInactive: bool = Query(False),
) -> list[dict[str, Any]]:
    """Every type a form may offer, in the administrator's order.

    NOT PAGINATED, AND THAT IS A DECISION. There are six rows; the screen that manages them and the
    dropdown that consumes them both want all of them, and a picker that could only see page one of
    its own vocabulary is the silent-truncation failure this repository keeps rediscovering. The
    ``@@index([isActive, sortOrder])`` serves this read exactly.

    SIGNED-IN AND OTHERWISE UNGATED for the same reason ``/reference`` is: what a form may offer is
    not per-user and is not sensitive, and a list a researcher could not read is a dropdown with no
    members. Writing is ``require_admin``; see the routes below.
    """
    where = {} if includeInactive else {"isActive": True}
    rows = await db.workshoptypeoption.find_many(where=where, order=_ORDER)
    return [_payload(row) for row in rows]


@router.post("", status_code=status.HTTP_201_CREATED)
async def create_workshop_type(
    payload: WorkshopTypeOptionCreate, _: Any = Depends(require_admin)
) -> dict[str, Any]:
    """Mint a type. The one request that decides a key.

    THE 409 IS THE UNIQUE INDEX ON ``key``, CONVERTED IN PLACE. Without it the driver's
    ``UniqueViolationError`` reaches the catch-all in ``main.py``, which logs a stack trace and
    answers 500 with "Something went wrong on the server" — telling an administrator the server is
    broken when the truth is that the token is taken and retrying cannot help. ``crafts.py`` converts
    the identical collision the identical way and its comment argues it at length.
    """
    try:
        created = await db.workshoptypeoption.create(data=payload.model_dump())
    except UniqueViolationError as exc:
        raise HTTPException(
            status_code=status.HTTP_409_CONFLICT,
            detail=(
                f"A workshop type with the key {payload.key} already exists. Keys are permanent "
                "because other tables store them, so edit that type's label rather than adding a "
                "second type under the same key."
            ),
        ) from exc
    return _payload(created)


@router.patch("/reorder")
async def reorder_workshop_types(
    payload: WorkshopTypeOptionReorder, _: Any = Depends(require_admin)
) -> list[dict[str, Any]]:
    """Set the order of the list in ONE request.

    ⚠ DECLARED BEFORE EVERY ``/{type_id}`` ROUTE, AND THAT POSITION IS LOAD-BEARING. FastAPI matches
    in registration order, and ``PATCH /{type_id}`` matches ``/reorder`` perfectly well — with
    ``type_id="reorder"``, which would answer 404 "Workshop type not found" on a server where this
    route exists. It is the same hazard ``api/router.py`` records for ``/design-workshops`` and its
    literal paths, one level down. Do not move this below the parameterised routes.

    ONE TRANSACTION, because a reorder that half-applied would leave the list in an order nobody
    chose — and, unlike a half-applied label edit, there is nothing on screen to say it happened.
    Positions are ``(index + 1) * 10``, leaving gaps so that a later insert-between does not have to
    renumber the world.

    AN UNKNOWN ID IS REFUSED, NOT SKIPPED. Skipping would renumber the rows around it and answer 200,
    so the administrator would see an order they did not ask for with nothing to say why.
    """
    rows = await db.workshoptypeoption.find_many()
    known = {row.id for row in rows}
    missing = [type_id for type_id in payload.ids if type_id not in known]
    if missing:
        raise HTTPException(
            status_code=status.HTTP_404_NOT_FOUND,
            detail=(
                "Cannot reorder: this server has no workshop type with "
                + ("id " if len(missing) == 1 else "ids ")
                + ", ".join(missing)
                + ". Reload the list — somebody else may have removed one."
            ),
        )
    if len(set(payload.ids)) != len(payload.ids):
        raise HTTPException(
            status_code=status.HTTP_400_BAD_REQUEST,
            detail="Cannot reorder: the same workshop type is named twice in one request.",
        )
    async with db.tx() as tx:
        for index, type_id in enumerate(payload.ids):
            await tx.workshoptypeoption.update(
                where={"id": type_id}, data={"sortOrder": (index + 1) * 10}
            )
    updated = await db.workshoptypeoption.find_many(order=_ORDER)
    return [_payload(row) for row in updated]


@router.patch("/{type_id}")
async def update_workshop_type(
    type_id: str, payload: WorkshopTypeOptionUpdate, _: Any = Depends(require_admin)
) -> dict[str, Any]:
    """Correct a type: its label, its position, whether it is offered, where it routes.

    **NEVER ITS KEY.** ``WorkshopTypeOptionUpdate`` does not carry the field and ``APIModel`` forbids
    extras, so a client that tries is answered 422 naming it. That is the guarantee the whole screen
    rests on — see the schema module's header for the four places the token is already stored.

    ``exclude_unset=True`` IS THE PRECONDITION OF THAT GUARANTEE and not a stylistic choice: it is
    what makes a present key mean "the caller sent this". Drop it and every optional the client left
    alone would arrive as ``None`` and be written as an explicit NULL over stored data — which on
    this model means a label edit would blank the routing flag.
    """
    row = await _require_type(type_id)
    data = payload.model_dump(exclude_unset=True)
    # An empty PATCH is a no-op, not an error: a form that sends only what changed legitimately sends
    # nothing when nothing changed, and answering 422 for it would make every client diff twice. The
    # row already read for the 404 is the answer — a second read to say the same thing would be a
    # round trip spent proving nothing happened.
    if not data:
        return _payload(row)
    updated = await db.workshoptypeoption.update(where={"id": type_id}, data=data)
    return _payload(updated)


@router.delete("/{type_id}", status_code=status.HTTP_204_NO_CONTENT)
async def delete_workshop_type(type_id: str, _: Any = Depends(require_admin)) -> None:
    """Delete a type that nothing is filed under. Refuse, with a count, when something is.

    See this module's header for the argument. The short version: there is no foreign key onto this
    table, so Postgres cannot make this refusal for us, and a delete that went through would leave
    every workshop filed under the key rendering a type nothing can resolve — indistinguishable from
    a workshop whose type was never answered.

    THE COUNT IS TAKEN INSIDE THE TRANSACTION THAT DELETES, so a workshop filed under this type
    between the count and the delete cannot slip through the gap. The count is two indexed reads; the
    gap it closes is small, and the cost of leaving it open is a type deleted out from under a
    workshop somebody was saving. Raising from inside the block rolls it back, which is the same
    arrangement ``crafts.py::update_craft`` uses to answer its 409 without leaving a half-written
    audit row behind.
    """
    row = await _require_type(type_id)
    async with db.tx() as tx:
        named, uncounted = await _usage_tally(tx, row.key)
        if named or uncounted:
            raise HTTPException(
                status_code=status.HTTP_409_CONFLICT,
                detail=_in_use_message(row.label, named, uncounted),
            )
        await tx.workshoptypeoption.delete(where={"id": type_id})

File: frontend/lib/workshopTypes.ts
/**
 * The "Type of workshop" vocabulary — the FIRST of the two dropdowns on every record form, and the
 * flag that decides where the SECOND one's answer is saved.
 *
 * ── THE NAME. IT IS `WorkshopTypeOption`, AND `WorkshopType` IS ALREADY SOMETHING ELSE ────────────
 *
 * `frontend/lib/types.ts` already exports `type WorkshopType = "DESIGN_PROTOTYPE" | "OTHER"`. That
 * is the two-member legacy enum on a `Workshop` ROW — it says whether one particular legacy workshop
 * was a design & prototype visit — and it is not this. This file's rows are the ADMINISTRATOR'S LIST
 * of the types a form may offer, and the picker in the record form holds BOTH meanings at the same
 * time, so a shared spelling would mean an import alias at every call site that touches either.
 *
 * The backend could not use the obvious name even if the client could: `WorkshopType` is a real
 * Postgres enum type on this database, and Postgres keeps enums and tables in one type namespace, so
 * `CREATE TABLE "WorkshopType"` answers `ERROR: type "WorkshopType" already exists`. See
 * `model WorkshopTypeOption` in `backend/prisma/schema.prisma`, which records the verification.
 *
 * ── WHAT A ROW IS FOR, AND THE ONE FIELD THAT MATTERS MOST ───────────────────────────────────────
 *
 * Two dropdowns, never three:
 *
 *     "Type of workshop"  → this list
 *     "Workshop"          → the workshops OF THAT TYPE, most recent first
 *
 * {@link WorkshopTypeOption.routesToDesignWorkshop} is what makes the second one possible without a
 * third control. TRUE means the workshop list comes from `DesignWorkshop` and the chosen row is
 * saved to the record's `designWorkshopId`; FALSE means it comes from `Workshop` and the chosen row
 * is saved to `workshopId`. One picker, two destinations, decided by the type.
 *
 * **THE TYPE ITSELF IS NOT SAVED ON THE RECORD.** The workshop the researcher picked already knows
 * its own type, so storing the type beside it would be a second copy that can disagree with the
 * first — and nothing would ever read them together to notice. The control is a NARROWING CONTROL
 * and nothing more — which is exactly what the retired `DesignWorkshopCascade` said of the filter
 * this replaces (that file was deleted on 2026-09-16; its hint read "Narrows the workshops below.
 * It is not saved on this record.").
 *
 * ── THIS IS NOT `stage_schema.ENUMS["WORKSHOP_KIND"]`, THOUGH THE KEYS ARE SEEDED FROM IT ────────
 *
 * `WORKSHOP_KIND` answers "what kind of design workshop is THIS one?" — a question inside stage 1 of
 * a design workshop's own questionnaire, answered by the designer running it. This list answers
 * "which lists may a record form offer?" — a question about the shape of a form, on six screens that
 * have no stage 1 at all. The six rows are SEEDED with the registry's six keys so that every
 * `DesignWorkshop.workshopKind` value already stored resolves to a label, and nothing synchronises
 * them after that. Do not derive one from the other in this client.
 */

import { apiFetch, buildQuery } from "@/lib/api";

/**
 * The key of the one seeded type that routes at `DesignWorkshop`.
 *
 * **READ IT, DO NOT COMPARE AGAINST IT.** This constant exists so a message can name the type in
 * words and so a test can assert the seed — it is NOT how a caller decides where to save. That
 * decision is {@link WorkshopTypeOption.routesToDesignWorkshop}, per row, because an administrator
 * may add a second design-workshop-backed programme the day the ministry announces one and no
 * client should have to ship to learn about it. A `key === DESIGN_PROTOTYPE_WORKSHOP_TYPE_KEY` test
 * in a picker is the bug this sentence exists to prevent.
 */
export const DESIGN_PROTOTYPE_WORKSHOP_TYPE_KEY = "DESIGN_PROTOTYPE_DEVELOPMENT";

/** One option in the "Type of workshop" dropdown, as `GET /workshop-types` serves it. */
export type WorkshopTypeOption = {
  id: string;
  /**
   * The stable token — `DESIGN_PROTOTYPE_DEVELOPMENT`, `SKILL_UPGRADATION`, … — and the value
   * `DesignWorkshop.workshopKind` and `AnnualPlanEntry.workshopKind` already store.
   *
   * **PERMANENT.** `PATCH /workshop-types/{id}` does not accept this field; the API answers 422 if a
   * client sends it. Editing a label is the common case and must never be able to move the token
   * that four other places already hold.
   */
  key: string;
  /** What a human reads in the dropdown. This is the field an administrator edits. */
  label: string;
  /** Low first. Not unique — `key` breaks the tie, which is what makes the order total. */
  sortOrder: number;
  /** False = retired: absent from every picker, with every workshop filed under it untouched. */
  isActive: boolean;
  /**
   * TRUE  → the "Workshop" dropdown is filled from `DesignWorkshop`, saved to `designWorkshopId`.
   * FALSE → it is filled from `Workshop`, saved to `workshopId`.
   *
   * Exactly one seeded row carries it, and no database constraint enforces "exactly one" — see the
   * Prisma model for why making that a constraint would make announcing a second programme a
   * migration on every deployment.
   */
  routesToDesignWorkshop: boolean;
  createdAt: string | null;
  updatedAt: string | null;
};

/* ────────────────────────────────────────────────────────────────────────────
 * Reading
 * ──────────────────────────────────────────────────────────────────────────── */

/**
 * Every type a form may offer, in the administrator's order.
 *
 * `includeInactive` IS FOR THE ADMIN SCREEN AND FOR NOTHING ELSE. A picker must never pass it: a
 * retired type absent from the dropdown is the entire meaning of retiring one. The list is not
 * paginated — there are six rows and the screen that manages them wants all of them — so there is no
 * truncation to state and no cap to print.
 */
export async function listWorkshopTypes(options?: { includeInactive?: boolean }) {
  const query = buildQuery({ includeInactive: options?.includeInactive ? "true" : undefined });
  return apiFetch<WorkshopTypeOption[]>(`/workshop-types${query}`);
}

/**
 * The server's own order, re-applied in the browser.
 *
 * Every consumer of a list in this client re-sorts rather than trusting the response order, and the
 * reason here is the ordinary one plus a specific one: `sortOrder` is deliberately NOT unique, so
 * two rows can share a position, and a caller that merged two responses (an admin screen that just
 * created a row, say) would otherwise render an order that depends on which array came first.
 * `key` breaks the tie exactly as the API's `ORDER BY "sortOrder", "key"` does.
 */
export function sortWorkshopTypes(types: WorkshopTypeOption[]): WorkshopTypeOption[] {
  return [...types].sort((a, b) => a.sortOrder - b.sortOrder || a.key.localeCompare(b.key));
}

/**
 * The type whose workshops live in `DesignWorkshop`, or null when the list holds none.
 *
 * READS THE FLAG, NEVER THE KEY — see {@link DESIGN_PROTOTYPE_WORKSHOP_TYPE_KEY}. Returns the FIRST
 * such type in the administrator's order rather than asserting there is one: nothing in the database
 * enforces "exactly one", so a caller that indexed `[0]` of a filter would be making a claim this
 * client cannot check, and a caller that threw would take out every record form the moment an
 * administrator added a second one.
 *
 * NULL IS A REAL ANSWER AND MUST BE HANDLED. An administrator can deactivate the design & prototype
 * type, and then a picker asking this question gets nothing back — which is correct, and means the
 * form offers only the ordinary-`Workshop` types until somebody turns it back on.
 */
export function designWorkshopTypeOf(types: WorkshopTypeOption[]): WorkshopTypeOption | null {
  return sortWorkshopTypes(types).find((type) => type.routesToDesignWorkshop) ?? null;
}

/**
 * Which type a form should open on.
 *
 * `preferDesignWorkshops` IS A CAPABILITY THE CALLER PASSES, NOT A ROLE THIS FILE READS. The rule is
 * "a designer opens on Design & Prototype", and the predicate for that lives in `lib/permissions.ts`
 * where every other one does; this file would otherwise be a second place that decides what a
 * designer is, and the two would disagree the first time the ladder moved. Pass
 * `canRunDesignWorkshops(user)`.
 *
 * FALLS BACK TO THE FIRST TYPE IN THE LIST rather than to null whenever it can, because the control
 * is not optional — a form that opened with no type selected would show an empty "Workshop"
 * dropdown, which reads as "there are no workshops" rather than as "choose a type first". Null comes
 * back only when there are no types at all, which is a state the seed makes unreachable except on a
 * database where an administrator has deactivated every one of them.
 */
export function defaultWorkshopType(
  types: WorkshopTypeOption[],
  options?: { preferDesignWorkshops?: boolean }
): WorkshopTypeOption | null {
  const ordered = sortWorkshopTypes(types.filter((type) => type.isActive));
  if (options?.preferDesignWorkshops) {
    const design = ordered.find((type) => type.routesToDesignWorkshop);
    if (design) return design;
  }
  return ordered[0] ?? null;
}

/**
 * The label for a stored token, for a screen that has a `workshopKind` and needs a word for it.
 *
 * FALLS BACK TO THE RAW TOKEN, NEVER TO "Unknown" AND NEVER TO AN EMPTY STRING. A workshop filed
 * under a type an administrator has since deleted — or one written by a client a release ahead — is
 * still a workshop with a type, and printing the token is the honest rendering of "this is what it
 * says". `enum_label` on the server falls back the same way and for the same reason.
 */
export function workshopTypeLabel(types: WorkshopTypeOption[], key: string | null | undefined) {
  if (!key) return null;
  return types.find((type) => type.key === key)?.label ?? key;
}

/* ────────────────────────────────────────────────────────────────────────────
 * Writing — ADMIN ONLY. Every one of these is `require_admin` on the server.
 * ──────────────────────────────────────────────────────────────────────────── */

/** `POST /workshop-types`. 409 when the key is taken — keys are permanent, so there is no retry. */
export async function createWorkshopType(body: {
  key: string;
  label: string;
  sortOrder?: number;
  isActive?: boolean;
  routesToDesignWorkshop?: boolean;
}) {
  return apiFetch<WorkshopTypeOption>("/workshop-types", { method: "POST", body: JSON.stringify(body) });
}

/**
 * `PATCH /workshop-types/{id}` — **send only what changed.**
 *
 * The server applies this with `exclude_unset`, so an absent key keeps the stored value. There is no
 * `key` in the body type and there must not be one: the API forbids extras and answers 422, which is
 * the guarantee that a label edit can never move a token four other tables store.
 */
export async function updateWorkshopType(
  id: string,
  body: { label?: string; sortOrder?: number; isActive?: boolean; routesToDesignWorkshop?: boolean }
) {
  return apiFetch<WorkshopTypeOption>(`/workshop-types/${id}`, {
    method: "PATCH",
    body: JSON.stringify(body)
  });
}

/**
 * `PATCH /workshop-types/reorder` — the whole order in one request, as a list of ids.
 *
 * ONE REQUEST AND NOT N PATCHES, because a reorder expressed as six saves is six chances to be
 * interrupted halfway and leave the list in an order nobody chose. The position IS the index in the
 * array, so a client cannot send an order that contradicts itself. Returns the reordered list.
 */
export async function reorderWorkshopTypes(ids: string[]) {
  return apiFetch<WorkshopTypeOption[]>("/workshop-types/reorder", {
    method: "PATCH",
    body: JSON.stringify({ ids })
  });
}

/**
 * `DELETE /workshop-types/{id}` — **and it is refused, with a count, when anything is filed under
 * the type.**
 *
 * 204 when nothing uses it. 409 otherwise, and the `detail` is a finished sentence naming how many
 * workshops stand in the way and telling the administrator to deactivate instead. RENDER THAT
 * SENTENCE VERBATIM: it is the only thing on screen that says what to do next, and a caller that
 * replaced it with "Unable to delete" would leave an administrator trying again tomorrow.
 *
 * The refusal exists because there is no foreign key onto the type table — a record stores the
 * WORKSHOP, never the type — so Postgres cannot make this refusal for us, and a delete that went
 * through would leave every workshop under that key rendering a type nothing can resolve.
 */
export async function deleteWorkshopType(id: string) {
  await apiFetch<void>(`/workshop-types/${id}`, { method: "DELETE" });
}
